File: LeetCode/Python Solutions/Remove Nth Node From End of List copy/remove.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # Initialize a dummy node to simplify edge cases like removing the head
        dummy = ListNode(0, head)
        slow = fast = dummy
        
        # Advance fast pointer n steps ahead
        for _ in range(n + 1):
            fast = fast.next
            
        # Move both pointers until fast reaches the end
        while fast:
            slow = slow.next
            fast = fast.next
            
        # Skip the target node
        slow.next = slow.next.next
        
        return dummy.next  # Return head of the modified list

File: LeetCode/Java Solutions/Remove Nth Node From End of List/remove.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // Initialize a dummy node to simplify edge cases like removing the head
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode slow = dummy, fast = dummy;
        
        // Advance fast pointer n steps ahead
        for (int i = 0; i < n + 1; i++) {
            fast = fast.next;
        }
        
        // Move both pointers until fast reaches the end
        while (fast != null) {
            slow = slow.next;
            fast = fast.next;
        }
        
        // Skip the target node
        slow.next = slow.next.next;
        
        return dummy.next;  // Return head of the modified list
    }
}